From 8703e8afb2882b6c3fa1096b4b13041589b050c1 Mon Sep 17 00:00:00 2001
From: Simon Josefsson <simon@josefsson.org>
Date: Mon, 25 Apr 2011 21:11:49 +0000
Subject: Support TOTPs in usersfile code.

Inspired by Rien Broekstra <rien@rename-it.nl> and Max Thoursie <max@maxnet.se>.
---
diff --git a/liboath/tests/Makefile.am b/liboath/tests/Makefile.am
index 8230c99..fc3bc1f 100644
--- a/liboath/tests/Makefile.am
+++ b/liboath/tests/Makefile.am
@@ -23,25 +23,25 @@ AM_LDFLAGS = -no-install
 
 LDADD = ../liboath.la
 
-EXTRA_DIST = users.hotp expect.hotp
+EXTRA_DIST = users.oath expect.oath
 
 TESTS = \
 	tst_basic$(EXEEXT) \
 	tst_errors$(EXEEXT) \
 	tst_hotp_algo$(EXEEXT) \
 	tst_hotp_validate$(EXEEXT) \
-	tst_hotp_usersfile.sh \
 	tst_totp_algo$(EXEEXT) \
-	tst_totp_validate$(EXEEXT)
+	tst_totp_validate$(EXEEXT) \
+	tst_usersfile.sh
 check_PROGRAMS = \
 	tst_basic$(EXEEXT) \
 	tst_errors$(EXEEXT) \
 	tst_hotp_algo$(EXEEXT) \
 	tst_hotp_validate$(EXEEXT) \
-	tst_hotp_usersfile$(EXEEXT) \
 	tst_totp_algo$(EXEEXT) \
-	tst_totp_validate$(EXEEXT)
-dist_check_SCRIPTS = tst_hotp_usersfile.sh
+	tst_totp_validate$(EXEEXT) \
+	tst_usersfile$(EXEEXT)
+dist_check_SCRIPTS = tst_usersfile.sh
 
 TESTS_ENVIRONMENT = EXEEXT=$(EXEEXT)
 
diff --git a/liboath/tests/expect.hotp b/liboath/tests/expect.oath
index 8aaa718..bf04d8b 100644
--- a/liboath/tests/expect.hotp
+++ b/liboath/tests/expect.oath
@@ -5,3 +5,4 @@ HOTP/E/8	joe	4711	01
 HOTP/E/6	jas	1234	3132333435363738393031323334353637383930
 HOTP/E/7	rms	6767	3132333435363738393031323334353637383930	15	436521	2006-12-07T00:00:00L
 HOTP		foo	8989	3132333435363738393031323334353637383930	0	755224	2009-12-07T17:25:42L
+HOTP/T30	eve	-	00	1	299833	2006-12-07T00:00:00L
diff --git a/liboath/tests/tst_hotp_usersfile.c b/liboath/tests/tst_usersfile.c
index 4d80329..13993d2 100644
--- a/liboath/tests/tst_hotp_usersfile.c
+++ b/liboath/tests/tst_usersfile.c
@@ -1,5 +1,5 @@
 /*
- * tst_hotp_usersfile.c - self-tests for Liboath usersfile functions
+ * tst_usersfile.c - self-tests for Liboath usersfile functions
  * Copyright (C) 2009-2011 Simon Josefsson
  *
  * This library is free software; you can redistribute it and/or
@@ -101,6 +101,30 @@ main (void)
       return 1;
     }
 
+  rc = oath_authenticate_usersfile (CREDS,
+				    "eve", "386397", 0, "4711", &last_otp);
+  if (rc != OATH_BAD_PASSWORD)
+    {
+      printf ("oath_authenticate_usersfile: %d\n", rc);
+      return 1;
+    }
+
+  rc = oath_authenticate_usersfile (CREDS,
+				    "eve", "299833", 0, NULL, &last_otp);
+  if (rc != OATH_INVALID_OTP)
+    {
+      printf ("oath_authenticate_usersfile: %d\n", rc);
+      return 1;
+    }
+
+  rc = oath_authenticate_usersfile (CREDS,
+				    "eve", "299833", 1, NULL, &last_otp);
+  if (rc != OATH_OK)
+    {
+      printf ("oath_authenticate_usersfile: %d\n", rc);
+      return 1;
+    }
+
   rc = oath_done ();
   if (rc != OATH_OK)
     {
diff --git a/liboath/tests/tst_hotp_usersfile.sh b/liboath/tests/tst_usersfile.sh
index 42d6ef6..567b820 100755
--- a/liboath/tests/tst_hotp_usersfile.sh
+++ b/liboath/tests/tst_usersfile.sh
@@ -1,5 +1,5 @@
 #!/bin/sh
-# tst_hotp_usersfile.sh - Invoke tst_hotp_usersfile and check output.
+# tst_usersfile.sh - Invoke tst_usersfile and check output.
 # Copyright (C) 2009-2011 Simon Josefsson
 
 # This library is free software; you can redistribute it and/or modify
@@ -17,19 +17,19 @@
 # Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
 # 02110-1301 USA
 
-# Check for datefudge
+cp $srcdir/users.oath tmp.oath
+
 TSTAMP=`datefudge "2006-09-23" date -u +%s`
 if test "$TSTAMP" != "1158962400"; then
     echo "Cannot fake timestamps, please install datefudge to check better."
-fi
-
-cp $srcdir/users.hotp tmp.oath
-
-if test "$TSTAMP" = "1158962400"; then
-    datefudge 2006-12-07 ./tst_hotp_usersfile$EXEEXT
-    diff -ur $srcdir/expect.hotp tmp.oath || exit 1
+    ./tst_usersfile$EXEEXT
+    rc=$?
 else
-    ./tst_hotp_usersfile$EXEEXT
+    datefudge 2006-12-07 ./tst_usersfile$EXEEXT
+    rc=$?
+    diff -ur $srcdir/expect.oath tmp.oath || rc=1
 fi
 
 rm -f tmp.oath
+
+exit $rc
diff --git a/liboath/tests/users.hotp b/liboath/tests/users.oath
index 6952a0e..92e4768 100644
--- a/liboath/tests/users.hotp
+++ b/liboath/tests/users.oath
@@ -5,3 +5,4 @@ HOTP/E/8	joe	4711	01
 HOTP/E/6	jas	1234	3132333435363738393031323334353637383930
 HOTP/E/7	rms	6767	3132333435363738393031323334353637383930	10
 HOTP		foo	8989	3132333435363738393031323334353637383930	0	755224	2009-12-07T17:25:42L
+HOTP/T30	eve	-	00
diff --git a/liboath/usersfile.c b/liboath/usersfile.c
index 8d9ce9a..8b99f59 100644
--- a/liboath/usersfile.c
+++ b/liboath/usersfile.c
@@ -31,19 +31,47 @@
 #include <errno.h>		/* For errno. */
 #include <sys/stat.h>		/* For S_IRUSR, S_IWUSR. */
 
-static unsigned
-parse_type (const char *str)
+static int
+parse_type (const char *str, unsigned *digits, unsigned *totpstepsize)
 {
-  if (strcmp (str, "HOTP/E/6") == 0)
-    return 6;
-  if (strcmp (str, "HOTP/E/7") == 0)
-    return 7;
-  if (strcmp (str, "HOTP/E/8") == 0)
-    return 8;
-  if (strcmp (str, "HOTP/E") == 0)
-    return 6;
-  if (strcmp (str, "HOTP") == 0)
-    return 6;
+  *totpstepsize = 0;
+  if (strcmp (str, "HOTP/E/6") == 0
+      || strcmp (str, "HOTP/E") == 0
+      || strcmp (str, "HOTP") == 0)
+    *digits = 6;
+  else if (strcmp (str, "HOTP/E/7") == 0)
+    *digits = 7;
+  else if (strcmp (str, "HOTP/E/8") == 0)
+    *digits = 8;
+  else if (strncmp (str, "HOTP/T30", 8) == 0)
+    {
+      *totpstepsize = 30;
+      if (strcmp (str, "HOTP/T30") == 0
+	  || strcmp (str, "HOTP/T30/6") == 0)
+	*digits = 6;
+      else if (strcmp (str, "HOTP/T30/7") == 0)
+	*digits = 7;
+      else if (strcmp (str, "HOTP/T30/8") == 0)
+	*digits = 8;
+      else
+	return -1;
+    }
+  else if (strncmp (str, "HOTP/T60", 8) == 0)
+    {
+      *totpstepsize = 60;
+      if (strcmp (str, "HOTP/T60") == 0
+	  || strcmp (str, "HOTP/T60/6") == 0)
+	*digits = 6;
+      else if (strcmp (str, "HOTP/T60/7") == 0)
+	*digits = 7;
+      else if (strcmp (str, "HOTP/T60/8") == 0)
+	*digits = 8;
+      else
+	return -1;
+    }
+  else
+    return -1;
+
   return 0;
 }
 
@@ -63,7 +91,7 @@ parse_usersfile (const char *username,
     {
       char *saveptr;
       char *p = strtok_r (*lineptr, whitespace, &saveptr);
-      unsigned digits;
+      unsigned digits, totpstepsize;
       char secret[32];
       size_t secret_length = sizeof (secret);
       uint64_t start_moving_factor = 0;
@@ -74,9 +102,8 @@ parse_usersfile (const char *username,
 	continue;
 
       /* Read token type */
-      digits = parse_type (p);
-      if (digits == 0)
-	continue;
+      if (parse_type (p, &digits, &totpstepsize) != 0)
+	  continue;
 
       /* Read username */
       p = strtok_r (NULL, whitespace, &saveptr);
@@ -142,8 +169,12 @@ parse_usersfile (const char *username,
       if (prev_otp && strcmp (prev_otp, otp) == 0)
 	return OATH_REPLAYED_OTP;
 
-      rc = oath_hotp_validate (secret, secret_length,
-			       start_moving_factor, window, otp);
+      if (totpstepsize == 0)
+	rc = oath_hotp_validate (secret, secret_length,
+				 start_moving_factor, window, otp);
+      else
+	rc = oath_totp_validate (secret, secret_length,
+				 time (NULL), totpstepsize, 0, window, otp);
       if (rc < 0)
 	return rc;
       *new_moving_factor = start_moving_factor + rc;
--
cgit v0.8.3.4
